pkg/utils/subscription.go

```diff
@@ -1267,6 +1303,19 @@ func DetectPlacementDecision(ctx context.Context, clReader client.Reader, clt cl
 	}
 }
 
+// DetectSubscription - Detect the Subscription API every 10 seconds. The controller will auto restart when CRs appear later.
+//
+//nolint:unparam
+func DetectSubscription(ctx context.Context, clReader client.Reader, clt client.Client) {
+	if !IsReadySubscription(clReader) {
+		go wait.UntilWithContext(ctx, func(ctx context.Context) {
+			if IsReadySubscription(clReader) {
+				os.Exit(1)
+			}
+		}, time.Duration(10)*time.Second)
+	}
+}
+
 func GetClientConfigFromKubeConfig(kubeconfigFile string) (*rest.Config, error) {
```

Comment:
Yes if the subscription APIs are not ready the Subscription controller will indeed still be running with errors, but that would be the same for trying to deploy a subscription at this time and once the APIs are ready the controller will terminate and start up again. You shouldn't be able to deploy a subscription during this time anyway as it wouldn't just be the controller that is missing the CRDs, but the whole hub. I will try to do it this other way as it would avoid the issue entirely.

Reply:
@xiangjingli I tried something like the suggested code and pushed my changes, but there is an issue that is occurring on the managed cluster in the addon-manager pod:
```
E0523 20:19:26.820022       1 manager.go:306] Subscription API is NOT ready: customresourcedefinitions.apiextensions.k8s.io "subscriptions.apps.open-cluster-management.io" is forbidden: User "system:open-cluster-management:cluster:cluster1:addon:application-manager:agent:application-manager" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope

```

Should I be doing the CR check on the managed clusters differently or do we just have to update the permissions?

Reply:
@lennysgarage  Yeah, I think doing the CR check is easier. So you may check the function IsReadySubscription in the for loop, sth like this:

```
for {
    isHubCluster := false
    if Options.ClusterName == "" {
        isHubCluster = true
    } 
    
    if  !utils.IsReadySubscription(sig, mgr.GetAPIReader(), mgr.GetClient(), isHubCluster) {
	klog.Error("Subscription API is NOT ready: ", err)
	time.Sleep(10 * time.Second)
    } else {
	klog.Info("Subscription API is ready.")
	break
    }
}
```

Reply:
@xiangjingli  Okay so I did some tests for both the `ApiextensionsV1().CustomResourceDefinitions().Get` way and putting the `IsReadySubscription` func in the for loop, and found that they actually don't solve the issue. While they wait for the Subscription CRDs to come online, and don't start the controller until they do, once they start the controller it still can't find the CRDs. The only time this actually fully worked was when we did the `IsReadySubscription` func in the goroutine because it would exit the controller after finding the CRDs allowing it to restart. Looks like kubernetes requires all the CRDs to exist before it starts up at all?